File: backend/app/collision.py
from __future__ import annotations

from .config import get_settings
from .schemas import HapticCommand, Risk, SceneObject


def _intensity(distance_m: float, warn: float, stop: float) -> int:
    if distance_m >= warn:
        return 0
    # Linear ramp from 90 at the warning edge to 255 at the stop threshold.
    span = max(warn - stop, 0.1)
    t = min(max((warn - distance_m) / span, 0.0), 1.0)
    return int(90 + t * 165)


def assess(objects: list[SceneObject]) -> tuple[Risk, HapticCommand]:
    """Map the nearest obstacle on each side to directional haptic intensity."""
    s = get_settings()
    haptic = HapticCommand()
    nearest = float("inf")

    for o in objects:
        level = _intensity(o.distance_m, s.collision_warn_m, s.collision_stop_m)
        if level == 0:
            continue
        nearest = min(nearest, o.distance_m)
        if o.side == "left":
            haptic.left = max(haptic.left, level)
        elif o.side == "right":
            haptic.right = max(haptic.right, level)
        else:
            haptic.center = max(haptic.center, level)

    if nearest <= s.collision_stop_m:
        risk: Risk = "stop"
        haptic.pattern = 2
    elif nearest < s.collision_warn_m:
        risk = "caution"
        haptic.pattern = 1
    else:
        risk = "clear"
        haptic.pattern = 0
    return risk, haptic

File: frontend/src/collision.js
// Collision assessment -> directional haptic command + Web Vibration pattern. Runs on edge.

function intensity(distance, warn, stop) {
  if (distance >= warn) return 0;
  const span = Math.max(warn - stop, 0.1);
  const t = Math.min(Math.max((warn - distance) / span, 0), 1);
  return Math.round(90 + t * 165);
}

export function assess(objects, warn, stop) {
  const haptic = { left: 0, center: 0, right: 0, pattern: 0 };
  let nearest = Infinity;

  for (const o of objects) {
    const level = intensity(o.distance_m, warn, stop);
    if (level === 0) continue;
    nearest = Math.min(nearest, o.distance_m);
    haptic[o.side] = Math.max(haptic[o.side], level);
  }

  let risk;
  if (nearest <= stop) { risk = "stop"; haptic.pattern = 2; }
  else if (nearest < warn) { risk = "caution"; haptic.pattern = 1; }
  else { risk = "clear"; haptic.pattern = 0; }
  return { risk, haptic };
}

export function vibrate(haptic) {
  if (!navigator.vibrate) return;
  const peak = Math.max(haptic.left, haptic.center, haptic.right);
  if (peak === 0) return;
  if (haptic.pattern === 2) navigator.vibrate([120, 60, 120, 60, 120]);
  else if (haptic.pattern === 1) navigator.vibrate([200, 120, 200]);
  else navigator.vibrate(80 + peak);
}
